Add Pop support to wwwAuthenticateParameters

trwalke opened the pull request:
Fixes # 
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/3026

**Changes proposed in this request**
Adding POP support to wwwAuthenticateParameters 
Revising api to support multiple WWW-Authenticate headers.
Adding support for Authentication-Info headers

Obsolete apis can be replaced with apis in the column on the immediate right except for `CreateFromWwwAuthenticateHeaderValue(string wwwAuthenticateValue)` which is completely deprecated. Developers should use `CreateFromAuthenticationHeaders(HttpResponseHeaders httpResponseHeaders)` instead.

| Obsolete APIs                                                                                                             | New APIs for Multiple Schemes (returns list of parsed schemes)                                                        | New APIs for Single Schemes                                                                                                          |
|---------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------|
| CreateFromResourceResponseAsync(string resourceUri)                                                                       | CreateFromAuthenticationResponseAsync(string resourceUri)                                                             | CreateFromAuthenticationResponseAsync(string resourceUri, string scheme)                                                             |
| CreateFromResourceResponseAsync(string resourceUri, CancellationToken cancellationToken = default)                        | CreateFromAuthenticationResponseAsync(string resourceUri, CancellationToken cancellationToken)                        | CreateFromAuthenticationResponseAsync(string resourceUri, string scheme, CancellationToken cancellationToken)                        |
| CreateFromResourceResponseAsync(HttpClient httpClient, string resourceUri, CancellationToken cancellationToken = default) | CreateFromAuthenticationResponseAsync(string resourceUri, HttpClient httpClient, CancellationToken cancellationToken) | CreateFromAuthenticationResponseAsync(string resourceUri, string scheme, HttpClient httpClient, CancellationToken cancellationToken) |
| WwwAuthenticateParameters CreateFromResponseHeaders(HttpResponseHeaders httpResponseHeaders, string scheme = "Bearer")    | CreateFromAuthenticationHeaders(HttpResponseHeaders httpResponseHeaders)                                              | CreateFromAuthenticationHeaders(HttpResponseHeaders httpResponseHeaders, string scheme)                                              |
| CreateFromWwwAuthenticateHeaderValue(string wwwAuthenticateValue)                                                         |                                                                                                                       |                                                                                                                                      |



**Testing**
Unit testing
integration testing

[API review](https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview?path=/WWW-Authenticate/WWWAuthenticate.md&version=GCe40f3a35efb2145173fe2e01a529afaaa2141079&_a=preview)

**Performance impact**
<!-- Describe any applicable performance impact or performance testing done. -->

trwalke commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -130,17 +180,41 @@ public static Task<WwwAuthenticateParameters> CreateFromResourceResponseAsync(st
+        /// <param name="resourceUri">URI of the resource.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel operation.</param>
+        /// <returns>a list of WWW-Authenticate Parameters extracted from a response to the unauthenticated call.</returns>
+        public static async Task<IEnumerable<WwwAuthenticateParameters>> CreateKnownParametersFromResourceResponseAsync(string resourceUri, CancellationToken cancellationToken = default)
```
Grabs a parsed list of all of the known/supported authentication schemes. This is to help facilitate fallback scenarios when Pop is not supported.

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -21,6 +22,14 @@ namespace Microsoft.Identity.Client
     {
+        private static readonly ISet<string> s_knownAuthenticationSchemes = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+        static WwwAuthenticateParameters()
```
avoid using static ctor, they are difficult to reason with. You can initialize the HashSet at line 25.

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -89,16 +108,47 @@ public class WwwAuthenticateParameters
+
+        private static WwwAuthenticateParameters ParsePopParameters(HttpResponseHeaders httpResponseHeaders)
+        {
+            if (httpResponseHeaders.WwwAuthenticate != null)
```
Why is there no check similar to the one at line 129?

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -130,17 +180,39 @@ public static Task<WwwAuthenticateParameters> CreateFromResourceResponseAsync(st
+        /// <param name="resourceUri">URI of the resource.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel operation.</param>
+        /// <returns>a list of WWW-Authenticate Parameters extracted from a response to the unauthenticated call.</returns>
+        public static async Task<IEnumerable<WwwAuthenticateParameters>> CreateKnownParametersFromResourceResponseAsync(string resourceUri, CancellationToken cancellationToken = default)
```
I recommend against calling the RP like this. We will almost certainly be asked to allow injection of HttpClient etc. So this is not the place for this sort of functionality.

Can't we just have a `CreateFromResponseHeaders(HttpResponse)` which returns a dictionary of params, [scheme] -> [WwwAuthenticateParameters] for all known schemes? Or maybe even all schemes.

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -50,6 +59,16 @@ public class WwwAuthenticateParameters
+        /// <summary>
+        /// Server Nonce.
+        /// </summary>
+        public string ServerNonce { get; set; }
```
wondering if we should use inheritance here, i.e. POPWWWAuthenticateParameters?

jmprieur commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -50,6 +59,16 @@ public class WwwAuthenticateParameters
+        /// <summary>
+        /// Scheme.
+        /// </summary>
+        public string Scheme { get; set; } 
```
I would rename it `AuthenticationProtocol`. Scheme is overloaded (with ASP.NET Core authentication scheme)

bgavrilMS replied:
The spec mentions "auth-scheme" but I agree it's overloaded

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate#syntax

jmprieur replied:
If the spec mentiosn auth-scheme, let's use AuthScheme, but let's make sure the comment provides a link to the spec, and has the name protocol.

jmprieur commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -50,6 +59,16 @@ public class WwwAuthenticateParameters
         public string Error { get; set; }
 
+        /// <summary>
+        /// Scheme.
```
Authentication protocol (Pop, Bearer, ...)

jmprieur commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -231,6 +306,11 @@ internal static WwwAuthenticateParameters CreateWwwAuthenticateParameters(IDicti
                 wwwAuthenticateParameters.Error = value;
             }
 
+            if (values.TryGetValue("WWWAuthenticate: PoP nonce", out value))
```
There can be a lot of things between WWWAuthenticate: and Pop nonce?

jmprieur reviewed (commented):
I left a few comments

pmaytak commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -112,10 +157,22 @@ public static WwwAuthenticateParameters CreateFromWwwAuthenticateHeaderValue(str
+            if (s_knownAuthenticationSchemes.Contains(AuthValuesSplit[0]))
+            {
+                parameters = CoreHelpers.SplitWithQuotes(AuthValuesSplit[1], ',')
+                    .Select(v => ExtractKeyValuePair(v.Trim()))
```
nit: don't think you need to trim here since, ExtractKeyValuePair trims.

pmaytak commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -125,22 +182,49 @@ public static WwwAuthenticateParameters CreateFromWwwAuthenticateHeaderValue(str
         /// </summary>
         /// <param name="resourceUri">URI of the resource.</param>
         /// <returns>WWW-Authenticate Parameters extracted from response to the unauthenticated call.</returns>
+        [Obsolete]
```
Add a note why obsolete and what else to use.

bgavrilMS replied:
Do we need to Obsolete these? Can't we add a "scheme" input param to them which defaults to bearer?

trwalke replied:
This is what I did for the previous iteration but we had a discussion and you mentioned we should obsolete apis that are not in alignment for what we are trying to do (Which is return multiple headers). Seems like we should reconsider this approach.

pmaytak commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -167,6 +252,31 @@ public static async Task<WwwAuthenticateParameters> CreateFromResourceResponseAs
+        /// <param name="cancellationToken">The cancellation token to cancel operation.</param>
+        /// <param name="scheme">Authentication scheme. Default is Bearer.</param>
+        /// <returns>WWW-Authenticate Parameters extracted from response to the unauthenticated call.</returns>
+        public static async Task<IEnumerable<WwwAuthenticateParameters>> CreateAllFromResourceResponseAsync(HttpClient httpClient, string resourceUri, CancellationToken cancellationToken = default, string scheme = "Bearer")
```
Use `Constants.BearerAuthHeaderPrefix`

trwalke replied:
Actually, this scheme param is here in error, need to be remove

pmaytak commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -253,6 +273,76 @@ public void ExtractClaimChallengeFromHeader_IncorrectError_ReturnNull()
+                                                         default).ConfigureAwait(false);
+
+            //Assert
+            Assert.IsTrue(parameterList.FirstOrDefault().AuthScheme == Constants.PoPAuthHeaderPrefix);
```
Assert.AreEqual.

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -92,15 +111,41 @@ public class WwwAuthenticateParameters
             {
-                // TODO: add POP support
-                AuthenticationHeaderValue bearer = httpResponseHeaders.WwwAuthenticate.First(v => string.Equals(v.Scheme, scheme, StringComparison.OrdinalIgnoreCase));
+                AuthenticationHeaderValue bearer = httpResponseHeaders.WwwAuthenticate.First(v => string.Equals(v.Scheme, Constants.BearerAuthHeaderPrefix, StringComparison.OrdinalIgnoreCase));
```
"scheme" is not used.

trwalke replied:
Yeah, this is a bug, missed in refactoring

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -92,15 +111,41 @@ public class WwwAuthenticateParameters
             {
-                // TODO: add POP support
-                AuthenticationHeaderValue bearer = httpResponseHeaders.WwwAuthenticate.First(v => string.Equals(v.Scheme, scheme, StringComparison.OrdinalIgnoreCase));
+                AuthenticationHeaderValue bearer = httpResponseHeaders.WwwAuthenticate.First(v => string.Equals(v.Scheme, Constants.BearerAuthHeaderPrefix, StringComparison.OrdinalIgnoreCase));
```
Correct, this would throw. `First` will throw an exception if the condition is not met.

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -92,15 +111,41 @@ public class WwwAuthenticateParameters
+        /// <param name="httpResponseHeaders">HttpResponseHeaders.</param>
+        /// <returns>The parameters requested by the web API.</returns>
+        /// <remarks>Currently it only supports the Bearer scheme</remarks>
+        public static IEnumerable<WwwAuthenticateParameters> CreateAllFromResponseHeaders(
```
I'd keep the same name i.e. `CreateFromResponseHeaders`

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -92,15 +111,41 @@ public class WwwAuthenticateParameters
+        /// <param name="httpResponseHeaders">HttpResponseHeaders.</param>
+        /// <returns>The parameters requested by the web API.</returns>
+        /// <remarks>Currently it only supports the Bearer scheme</remarks>
+        public static IEnumerable<WwwAuthenticateParameters> CreateAllFromResponseHeaders(
```
how about we return `IReadOnlyList` ?

bgavrilMS replied:
Actually it looks like the .NET guidelines provide an answer here: 

- use IEnumerable<T> for input params
- use ReadOnlyCollection<T> for return types 

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/guidelines-for-collections?redirectedfrom=MSDN - very good and short read (CC @pmaytak , @jmprieur, @gladjohn, @neha-bhargava, @SameerK-MSFT  )

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -92,15 +111,41 @@ public class WwwAuthenticateParameters
             {
-                // TODO: add POP support
-                AuthenticationHeaderValue bearer = httpResponseHeaders.WwwAuthenticate.First(v => string.Equals(v.Scheme, scheme, StringComparison.OrdinalIgnoreCase));
+                AuthenticationHeaderValue bearer = httpResponseHeaders.WwwAuthenticate.First(v => string.Equals(v.Scheme, Constants.BearerAuthHeaderPrefix, StringComparison.OrdinalIgnoreCase));
```
I don't see why this method can't handle any authentication scheme out there. it's just parsing logic.

trwalke replied:
It can, this is just a bug left over from testing/refactoring

bgavrilMS commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -292,5 +382,17 @@ private static HttpResponseMessage CreateInvalidTokenHttpErrorResponse(string te
+            {
+                Headers =
+                {
+                    { WwwAuthenticateHeaderName, $"Bearer authorization_uri=\"https://login.microsoftonline.com/TenantId/oauth2/authorize\", resource_id=\"https://endpoint/\"" },
```
What about tests for the other sytnaxes? https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate#syntax

bgavrilMS commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -253,6 +273,76 @@ public void ExtractClaimChallengeFromHeader_IncorrectError_ReturnNull()
+        {
+            //Arrange & Act
+            var parameterList = await WwwAuthenticateParameters.CreateAllFromResourceResponseAsync(
+                                                         "https://testingsts.azurewebsites.net/servernonce/invalidsignature",
```
Is this hitting the network? If it is, it should be an integration test.

trwalke commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -92,15 +111,41 @@ public class WwwAuthenticateParameters
                 string wwwAuthenticateValue = bearer.Parameter;
-                return CreateFromWwwAuthenticateHeaderValue(wwwAuthenticateValue);
+                var parameters = CreateFromWwwAuthenticateHeaderValue(wwwAuthenticateValue);
+                parameters.AuthScheme = Constants.BearerAuthHeaderPrefix;
```
Yes, this should be doing that, its a bug.

trwalke commented:
next iteration will change `IEnumerable<WwwAuthenticateParameters>>` to `Dictionary<string, WwwAuthenticateParameters>>` where string is the scheme

bgavrilMS commented:
Converted to draft while waiting for design discussion to complete.

trwalke commented:
Addressed feedback in API review. needs one more look. @bgavrilMS @pmaytak @SameerK-MSFT @neha-bhargava @gladjohn 

https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview?path=/WWW-Authenticate/WWWAuthenticate.md&version=GCe40f3a35efb2145173fe2e01a529afaaa2141079&_a=preview

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -1,281 +0,0 @@
-    /// scenarios such as claim challenge, CAE, CA auth context.
-    /// See https://aka.ms/msal-net/wwwAuthenticate.
-    /// </summary>
-    public class WwwAuthenticateParameters
```
You can't delete a public class, it needs to be depreacated.

bgavrilMS commented on src/client/Microsoft.Identity.Client/Http/Headers/AuthenticationHeaderParser.cs:
```diff
@@ -0,0 +1,154 @@
+    /// <summary>
+    /// Parsed authentication headers to retreve header values from HttpResponseHeaders.
+    /// </summary>
+    public class AuthenticationHeaderParser
```
I think you should place this class in the root, where the `WWWAuthenticateParameteers` used to live.

bgavrilMS commented on src/client/Microsoft.Identity.Client/Http/Headers/WwwAuthenticateParameters.cs:
```diff
@@ -0,0 +1,464 @@
+        /// Resource for which to request scopes.
+        /// This is the App ID URI of the API that returned the WWW-Authenticate header.
+        /// </summary>
+        [Obsolete("The client apps should know which App ID URI it requests scopes for.", true)]
```
By moving it to a new location, you've made it impossible to review. Reviewers cannot see the changes brought to this file.

bgavrilMS commented on src/client/Microsoft.Identity.Client/Http/Headers/WwwAuthenticateParameters.cs:
```diff
@@ -0,0 +1,464 @@
+        /// <param name="resourceUri">URI of the resource.</param>
+        /// <returns>WWW-Authenticate Parameters extracted from response to the unauthenticated call.</returns>
+        [System.ComponentModel.EditorBrowsable(EditorBrowsableState.Never)]
+        public static Task<WwwAuthenticateParameters> CreateFromResourceResponseAsync(string resourceUri)
```
I feel that we should add an "Obsolete" attribute to these ones, because they are not really usable as they are.

trwalke replied:
Some solutions set the use of obsolete methods as an error in their ruleset, which is technically a breaking change. I believe the recommended SDL ruleset has this which can cause issues. Also, the obsolete methods still work, they just don't account for Pop/Multiple schemes. Ill add `<remarks>`

bgavrilMS commented on src/client/Microsoft.Identity.Client/Http/Headers/WwwAuthenticateParameters.cs:
```diff
@@ -0,0 +1,464 @@
+        /// Create the authenticate parameters by attempting to call the resource unauthenticated, and analyzing the response.
+        /// </summary>
+        /// <param name="resourceUri">URI of the resource.</param>
+        /// <param name="scheme">Authentication scheme.</param>
```
Give examples in the comments of schemes.

bgavrilMS commented on src/client/Microsoft.Identity.Client/Http/Headers/WwwAuthenticateParameters.cs:
```diff
@@ -0,0 +1,464 @@
+
+            var authValuesSplit = wwwAuthenticateValue.Split(new char[] { ' ' }, 2);
+
+            if (s_knownAuthenticationSchemes.Contains(authValuesSplit[0]))
```
I was under the impression that we are parsing all schemes? Not sure where the spec landed on this.

bgavrilMS replied:
`The apis that return dictionaries should try to return a collection of all of the WWW-AuthenticateHeaders found, not just the ones we know about.`

trwalke replied:
This is actually one of the obsolete apis. Also, this check for s_knownAuthenticationSchemes was used to check if the developer passed in the wwwAuthenticateValue with the auth scheme added to it which caused issues with parsing. The new apis parse the entire collection of WWW-Authentication headers.

bgavrilMS commented on src/client/Microsoft.Identity.Client/Http/Headers/WwwAuthenticateParameters.cs:
```diff
@@ -0,0 +1,464 @@
+            if (values.Count == 0)
+            {
+                //unable to parse auth header values
+                wwwAuthenticateParameters.ParsingError = true;
```
Add some context to this error via the "Error" property.

bgavrilMS commented on src/client/Microsoft.Identity.Client/Http/Headers/WwwAuthenticateParameters.cs:
```diff
@@ -0,0 +1,464 @@
+        /// <returns>a json fragment (original input string or decoded from base64 encoded).</returns>
+        private static string GetJsonFragment(string inputString)
+        {
+            if (string.IsNullOrEmpty(inputString) || inputString.Length % 4 != 0 || inputString.Any(c => char.IsWhiteSpace(c)))
```
I am not sure about this. How about: 

1. Try to base64url decode it. If this works and the result is JSON, use it.
2. Try to see if the result is JSON. If it is, use it.
3. Fail with parsing error?

trwalke replied:
Yeah, I didnt write this so not 100% sure what the checks are looking for. ill see if I can rework it.

bgavrilMS commented on src/client/Microsoft.Identity.Client/Http/Headers/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,57 @@
+            {
+                var authInfoValue = respnseHeaders.Where(header => header.Key == _authenticationInfoKey).Single().Value.FirstOrDefault();
+
+                var AuthValuesSplit = authInfoValue.Split(new char[] { ' ' }, 2);
```
naming.

bgavrilMS commented on src/client/Microsoft.Identity.Client/Http/Headers/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,57 @@
+        /// <summary>
+        /// 
+        /// </summary>
+        public string NextNonce { get; private set; }
```
Should this also have "Error" and "ParserError" ?

trwalke replied:
Yes, that makes sense

bgavrilMS commented on src/client/Microsoft.Identity.Client/Http/Headers/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,57 @@
+        /// </summary>
+        /// <param name="respnseHeaders"></param>
+        /// <returns></returns>
+        public static AuthenticationInfoParameters CreateFromHeaders(HttpResponseHeaders respnseHeaders)
```
typo in param name.

bgavrilMS commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -253,6 +286,60 @@ public void ExtractClaimChallengeFromHeader_IncorrectError_ReturnNull()
         }
 
+        [TestMethod]
+        public async Task CreateAllFromResourceResponseAsync_HttpClient_Bearer_Pop_Async()
```
Can you measure the code coverage on the new code? I would expect this to be fully (100%) covered.

bgavrilMS replied:
I did this and the results are good. I think the only untested block of code is 

<img width="889" alt="image" src="https://user-images.githubusercontent.com/12273384/207066752-4c74846b-4880-4e29-82a4-56196bee27c9.png">


To do this yourself, right click on the test class and select "analyze test coverage"

<img width="271" alt="image" src="https://user-images.githubusercontent.com/12273384/207066919-174b6dca-d463-4ea1-b5ec-e6df82c9cf68.png"> and then just navigate to the product code and you'll see it highlighted. Missing coverage is in red, good coverage in blue.

trwalke replied:
It is actually kinda difficult to create a test that throws an exception. Whenever I try to create a header with a malformed value, the .NET code throws an exception before the test method has a chance to execute. ill have to keep playing around with it.

bgavrilMS commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -156,6 +157,14 @@ public async Task CreateFromResourceResponseAsync_HttpClient_Arm_GetTenantId_Asy
             var authParams = await WwwAuthenticateParameters.CreateFromResourceResponseAsync(httpClient, resourceUri).ConfigureAwait(false);
 
             Assert.AreEqual(authParams.GetTenantId(), tenantId);
+
```
THere doesn't seem to be enough unit tests: 

- claims in base64url format / json format / bad format
- more parsing errors
- extra auth schemes.

bgavrilMS replied:
Can you get the tests from MSAL C++ and see if we can adapt them here?

bgavrilMS commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -292,5 +379,17 @@ private static HttpResponseMessage CreateInvalidTokenHttpErrorResponse(string te
+            {
+                Headers =
+                {
+                    { WwwAuthenticateHeaderName, $"Bearer authorization_uri=\"https://login.microsoftonline.com/TenantId/oauth2/authorize\", resource_id=\"https://endpoint/\"" },
```
The parser does not cover the case of `// Challenges specified in single header
WWW-Authenticate: challenge1, ..., challengeN`. This is probably ok, but an explicit error needs to be added.

trwalke replied:
This scenario is covered by the .Net Http Parsing logic. However, it work hurt t add the test case to be 100% sure.

bgavrilMS commented on src/client/Microsoft.Identity.Client/Http/Headers/WwwAuthenticateParameters.cs:
```diff
@@ -0,0 +1,464 @@
+
+            var authValuesSplit = wwwAuthenticateValue.Split(new char[] { ' ' }, 2);
+
+            if (s_knownAuthenticationSchemes.Contains(authValuesSplit[0]))
```
please use more variables, e.g. `authValuesSplit[1]`

bgavrilMS reviewed (changes requested):
Move files back. More tests needed.

neha-bhargava commented on src/client/Microsoft.Identity.Client/Http/Headers/AuthenticationHeaderParser.cs:
```diff
@@ -0,0 +1,154 @@
+        public IReadOnlyList<WwwAuthenticateParameters> ParsedWwwAuthenticateParameters { get; private set; }
+
+        /// <summary>
+        /// Parameters returned by the Authenticatin-Info header. 
```
```suggestion
        /// Parameters returned by the Authentication-Info header. 
```

neha-bhargava commented on src/client/Microsoft.Identity.Client/Http/Headers/AuthenticationHeaderParser.cs:
```diff
@@ -0,0 +1,154 @@
+        /// Creates a parsed set of parameters from the provided HttpResponseHeaders.
+        /// </summary>
+        /// <param name="httpResponseHeaders"></param>
+        /// <remarks>For known values, such as the nonce used for Proof-of-Possession, the parser will first chack for in the WWW-Authenticate headers
```
```suggestion
        /// <remarks>For known values, such as the nonce used for Proof-of-Possession, the parser will first check for it in the WWW-Authenticate headers
```

neha-bhargava commented on src/client/Microsoft.Identity.Client/Http/Headers/AuthenticationHeaderParser.cs:
```diff
@@ -0,0 +1,154 @@
+            }
+
+            //If no WWW-AuthenticateHeaders exist, attempt to parse AuthenticationInfo headers instead
+            AuthenticationInfoParameters authenticationInfoParameters = AuthenticationInfoParameters.CreateFromHeaders(httpResponseHeaders);
```
Should this be moved to the else block if this needs to be performed in case no www-authenticateHeaders are found?

pmaytak reviewed (changes requested):
Please run the ApiCompat tool before merge to verify intended public API changes.
Also the wwwAuth wiki needs to be updated. I think the code samples from the Api review spec are pretty helpful and should be included in the wiki.

pmaytak commented on src/client/Microsoft.Identity.Client/Http/Headers/WwwAuthenticateParameters.cs:
```diff
@@ -0,0 +1,464 @@
+            }
+            IDictionary<string, string> parameters;
+
+            var authValuesSplit = wwwAuthenticateValue.Split(new char[] { ' ' }, 2);
```
Per [spec](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate#syntax) the value can have multiple challenges (ex. bearer, pop) separated by comma. So we should split by `,` and this method should return a collection of WwwAuthenticateParameters really.

trwalke replied:
Actually, this method is also supposed to be obsolete. Instead of allowing developers to pass in any string they want, we are now going to only allow them to pass in the `Headers` collection from the .NET http response. .NET will automatically seperate headers by `'` and create a list of challenges that we can parse individually.

jmprieur commented:
Adding @brentschmaltz for his perspective

trwalke commented on src/client/Microsoft.Identity.Client/Http/Headers/WwwAuthenticateParameters.cs:
```diff
@@ -0,0 +1,464 @@
+        public string Claims { get; set; }
+
+        /// <summary>
+        /// Error.
```
Error represents what is returned from the resource if there is an error. It is returned in the WWW-Authenticate header

jmprieur commented on src/client/Microsoft.Identity.Client/Http/Headers/WwwAuthenticateParameters.cs:
```diff
@@ -0,0 +1,464 @@
+        /// <summary>
+        /// Error indicating that parsing failed.
+        /// </summary>
+        public bool ParsingError { get; private set; }
```
wouldn't it better to throw an exception if the header cannot be parsed?

trwalke replied:
Yeah, on second thought, an exception is better.

jmprieur commented on src/client/Microsoft.Identity.Client/Http/Headers/AuthenticationHeaderParser.cs:
```diff
@@ -0,0 +1,154 @@
+        /// <returns>Key Value pair</returns>
+        internal static KeyValuePair<string, string> ExtractKeyValuePair(string assignment)
+        {
+            string[] segments = CoreHelpers.SplitWithQuotes(assignment, '=')
```
@brentschmaltz saved a lot of CPU in Wilson by using spans between indexes.

trwalke commented on src/client/Microsoft.Identity.Client/Http/Headers/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,57 @@
+
+                string value;
+
+                if (paramValues.TryGetValue("nextnonce", out value))
```
from what I see in the spec, yes

trwalke commented on src/client/Microsoft.Identity.Client/Http/Headers/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,57 @@
+
+            if (respnseHeaders.Any(header => header.Key == _authenticationInfoKey))
+            {
+                var authInfoValue = respnseHeaders.Where(header => header.Key == _authenticationInfoKey).Single().Value.FirstOrDefault();
```
Yes, this is correct. It is similar to the WWW-Authenticate headers

trwalke replied:
I dont think there will be multiple `Authentication-Info` headers, rather, a list of comma separated values in the one `Authentication-Info`. I dont see anything in the spec suggesting there should be more than one. 
see https://www.rfc-editor.org/rfc/rfc7615

trwalke commented on src/client/Microsoft.Identity.Client/Http/Headers/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,57 @@
+
+            if (respnseHeaders.Any(header => header.Key == _authenticationInfoKey))
+            {
+                var authInfoValue = respnseHeaders.Where(header => header.Key == _authenticationInfoKey).Single().Value.FirstOrDefault();
```
We do a check on the line above to ensure the header will exist.

trwalke commented on src/client/Microsoft.Identity.Client/Http/Headers/AuthenticationHeaderParser.cs:
```diff
@@ -0,0 +1,154 @@
+            {
+                var WwwParameters = WwwAuthenticateParameters.CreateFromAuthenticateHeaders(httpResponseHeaders);
+
+                if (WwwParameters.Any(parameter => parameter.AuthScheme == "PoP"))
+                {
+                    serverNonce = WwwParameters.Where(parameter => parameter.AuthScheme == "PoP").Single().ServerNonce;
```
There is a check on the previous line to ensure the value exists

bgavrilMS commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -253,6 +358,131 @@ public void ExtractClaimChallengeFromHeader_IncorrectError_ReturnNull()
+        [TestMethod]
+        [DataRow(false)]
+        [DataRow(true)]
+        public void ExtractAllAuthinfoParametersFromResponseWithAuthParser(bool combineHeaders)
```
param is not used.

pmaytak commented on src/client/Microsoft.Identity.Client/AuthenticationHeaderParser.cs:
```diff
@@ -0,0 +1,155 @@
+        /// scenarios such as claim challenge, CAE, CA auth context.
+        /// See https://aka.ms/msal-net/wwwAuthenticate.
+        /// </summary>
+        public IReadOnlyList<WwwAuthenticateParameters> WwwAuthenticateParameters { get; private set; }
```
This can be defaulted to an empty list?

trwalke replied:
it set to empty lists by default in the ParseAuthenticationHeaders method.

pmaytak commented on src/client/Microsoft.Identity.Client/AuthenticationHeaderParser.cs:
```diff
@@ -0,0 +1,155 @@
+
+                if (wwwParameters.Any(parameter => parameter.AuthScheme == "PoP"))
+                {
+                    serverNonce = wwwParameters.Where(parameter => parameter.AuthScheme == "PoP").Single().Nonce;
```
```suggestion
                    serverNonce = wwwParameters.Single(parameter => parameter.AuthScheme == "PoP").Nonce;
```

pmaytak commented on src/client/Microsoft.Identity.Client/MsalErrorMessage.cs:
```diff
@@ -432,6 +432,8 @@ public static string CertMustHavePrivateKey(string certificateName)
 
         public const string RequestFailureErrorMessagePii = "=== Token Acquisition ({0}) failed:\n\tAuthority: {1}\n\tClientId: {2}.";
 
+        public const string UnableToParseAuthenticationHeader = "Msal is unable to parse the authentication reader returned from the endpoint";
```
Maybe more details here would be helpful for possible reasons why this may happen, like wwwAuth and auth-info headers missing, error parsing, etc?

pmaytak commented on src/client/Microsoft.Identity.Client/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,99 @@
+            {
+                if (httpResponseHeaders.Contains(AuthenticationInfoKey))
+                {
+                    var authInfoValue = httpResponseHeaders.Where(header => header.Key == AuthenticationInfoKey).Single().Value.FirstOrDefault();
```
```suggestion
                    var authInfoValue = httpResponseHeaders.Single(header => header.Key == AuthenticationInfoKey).Value.FirstOrDefault();
```

pmaytak commented on src/client/Microsoft.Identity.Client/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,99 @@
+                {
+                    var authInfoValue = httpResponseHeaders.Where(header => header.Key == AuthenticationInfoKey).Single().Value.FirstOrDefault();
+
+                    if (authInfoValue != null)
```
I don't think this will ever be not null, since the outer if Contains returns true. Then also we'd want to throw if Contains retuns false, right? So the try/catch could just wrap the parsing logic.

pmaytak commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -86,61 +155,61 @@ public class WwwAuthenticateParameters
-                AuthenticationHeaderValue bearer = httpResponseHeaders.WwwAuthenticate.First(v => string.Equals(v.Scheme, scheme, StringComparison.OrdinalIgnoreCase));
-                string wwwAuthenticateValue = bearer.Parameter;
-                return CreateFromWwwAuthenticateHeaderValue(wwwAuthenticateValue);
+                AuthenticationHeaderValue headerValue = httpResponseHeaders.WwwAuthenticate.FirstOrDefault(v => string.Equals(v.Scheme, Constants.BearerAuthHeaderPrefix, StringComparison.OrdinalIgnoreCase));
```
```suggestion
                AuthenticationHeaderValue headerValue = httpResponseHeaders.WwwAuthenticate.FirstOrDefault(v => string.Equals(v.Scheme, scheme, StringComparison.OrdinalIgnoreCase));
```

pmaytak commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -191,15 +297,154 @@ public static async Task<WwwAuthenticateParameters> CreateFromResourceResponseAs
+                    var parameters = CreateFromWwwAuthenticationHeaderValue(wwwAuthenticateHeaderValue.Parameter, wwwAuthenticateHeaderValue.Scheme);
+                    parameterList.Add(parameters);
+                }
+                catch (Exception ex)
+                {
+                    if (ex is MsalException)
+                    {
+                        throw;
+                    }
+
+                    throw new MsalClientException(MsalError.UnableToParseAuthenticationHeader, MsalErrorMessage.UnableToParseAuthenticationHeader, ex);
+                }
```
```suggestion
                catch (Exception ex) when (ex is not MsalException)
                {
                    throw new MsalClientException(MsalError.UnableToParseAuthenticationHeader, MsalErrorMessage.UnableToParseAuthenticationHeader, ex);
                }
```

gladjohn commented on src/client/Microsoft.Identity.Client/AuthenticationHeaderParser.cs:
```diff
@@ -0,0 +1,155 @@
+namespace Microsoft.Identity.Client
+{
+    /// <summary>
+    /// Parsed authentication headers to retreve header values from HttpResponseHeaders.
```
typo

gladjohn commented on src/client/Microsoft.Identity.Client/AuthenticationHeaderParser.cs:
```diff
@@ -0,0 +1,155 @@
+
+        /// <summary>
+        /// Parameters returned by the Authentication-Info header. 
+        /// This allows for authentication scenarios such as Proof-Of-Posession.
```
```suggestion
        /// This allows for authentication scenarios such as Proof-Of-Possession.
```

gladjohn commented on src/client/Microsoft.Identity.Client/AuthenticationHeaderParser.cs:
```diff
@@ -0,0 +1,155 @@
+        /// Creates the authenticate parameters by attempting to call the resource unauthenticated, and analyzing the response.
+        /// </summary>
+        /// <param name="resourceUri">URI of the resource.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel operation.</param>
```
```suggestion
        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
```

gladjohn commented on src/client/Microsoft.Identity.Client/AuthenticationHeaderParser.cs:
```diff
@@ -0,0 +1,155 @@
+        /// Creates the authenticate parameters by attempting to call the resource unauthenticated, and analyzing the response.
+        /// </summary>
+        /// <param name="resourceUri">URI of the resource.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel operation.</param>
```
nit : same as above

bgavrilMS commented on src/client/Microsoft.Identity.Client/AuthenticationHeaderParser.cs:
```diff
@@ -0,0 +1,155 @@
+        public AuthenticationInfoParameters AuthenticationInfoParameters { get; private set; }
+
+        /// <summary>
+        /// Nonce parsed from HttpResponseHeaders
```
You should mention that is only extracted from the POP WWW-Authenticate header

bgavrilMS commented on src/client/Microsoft.Identity.Client/AuthenticationHeaderParser.cs:
```diff
@@ -0,0 +1,155 @@
+        /// <summary>
+        /// Nonce parsed from HttpResponseHeaders
+        /// </summary>
+        public string Nonce { get; private set; }
```
I wonder if should call this PoPNonce ?

bgavrilMS commented on src/client/Microsoft.Identity.Client/AuthenticationHeaderParser.cs:
```diff
@@ -0,0 +1,155 @@
+        /// </summary>
+        internal static HttpClient GetHttpClient()
+        {
+            var httpClientFactory = PlatformProxyFactory.CreatePlatformProxy(null).CreateDefaultHttpClientFactory();
```
factory can be set only once (via Lazy) instead of created all the time.

bgavrilMS commented on src/client/Microsoft.Identity.Client/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,99 @@
+                    throw new MsalClientException(MsalError.UnableToParseAuthenticationHeader, MsalErrorMessage.UnableToParseAuthenticationHeader);
+                }
+
+                return parameters;
```
If there is no AuthenticationInfo header, shouldn't we return null?

bgavrilMS commented on src/client/Microsoft.Identity.Client/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,99 @@
+                    {
+                        var AuthValuesSplit = authInfoValue.Split(new char[] { ' ' }, 2);
+
+                        var paramValues = CoreHelpers.SplitWithQuotes(AuthValuesSplit[1], ',')
```
nit: we'll need to optimize this for perf (separately)

bgavrilMS commented on src/client/Microsoft.Identity.Client/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,99 @@
+                    }
+
+                    //Could not get Auth info parameters
+                    throw new MsalClientException(MsalError.UnableToParseAuthenticationHeader, MsalErrorMessage.UnableToParseAuthenticationHeader);
```
A more specific error message should be thrown.

bgavrilMS commented on src/client/Microsoft.Identity.Client/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,99 @@
+
+                return parameters;
+            }
+            catch(Exception ex)
```
nit: ` catch (Exception e) when !(e is MsalClientException)`

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/when

pmaytak replied:
You can use `not`. Pretty cool.
`catch (Exception ex) when (ex is not MsalClientException)`

bgavrilMS commented on src/client/Microsoft.Identity.Client/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,99 @@
+                {
+                    throw;
+                }
+                throw new MsalClientException(MsalError.UnableToParseAuthenticationHeader, MsalErrorMessage.UnableToParseAuthenticationHeader, ex);
```
nit: Add better error message, i.e. " ... see inner exception for details".

bgavrilMS commented on src/client/Microsoft.Identity.Client/AuthenticationHeaderParser.cs:
```diff
@@ -103,7 +103,7 @@ public static AuthenticationHeaderParser ParseAuthenticationHeaders(HttpResponse
                 if (wwwParameters.Any(parameter => parameter.AuthScheme == "PoP"))
                 {
-                    serverNonce = wwwParameters.Where(parameter => parameter.AuthScheme == "PoP").Single().Nonce;
+                    serverNonce = wwwParameters.Single(parameter => parameter.AuthScheme == "PoP").Nonce;
```
avoid double iteration here. Both `Any` and `Single` will iterate the collection to apply the condition. 

You can do 

```
var serverNonce = wwwParameters.SingleOrDefault(parameter => parameter.AuthScheme == "PoP")?.Nonce
```

bgavrilMS replied:
also protect against NPE, Nonce could be null.

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/WwwAuthenticateParametersIntegrationTests.cs:
```diff
@@ -60,5 +73,116 @@ public async Task CreateWwwAuthenticateResponseFromAzureResourceManagerUrlAsync(
+            await TestCommon.ValidatePopNonceAsync(parsedHeaders.Nonce).ConfigureAwait(false);
+        }
+
+#if NET_CORE
```
If it's not working, then there is no point if adding `#if NET_CORE`

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/WwwAuthenticateParametersIntegrationTests.cs:
```diff
@@ -60,5 +73,116 @@ public async Task CreateWwwAuthenticateResponseFromAzureResourceManagerUrlAsync(
+            IPublicClientApplication pca = PublicClientApplicationBuilder
+               .Create(labResponse.App.AppId)
+               .WithAuthority(labResponse.Lab.Authority, "organizations")
+               .WithExperimentalFeatures()
```
No longer needed.

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -50,6 +58,17 @@ public class WwwAuthenticateParameters
+        /// <summary>
+        /// Server Nonce.
+        /// </summary>
+        public string Nonce { get; private set; }
```
POP Nonce?

pmaytak replied:
👍 `PopNonce`

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -78,6 +97,54 @@ public class WwwAuthenticateParameters
         public string GetTenantId() => Instance.Authority
                                                .CreateAuthority(Authority, validateAuthority: true)
```
Possible null ref.

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -86,61 +153,61 @@ public class WwwAuthenticateParameters
-                .ToDictionary(pair => pair.Key, pair => pair.Value, StringComparer.OrdinalIgnoreCase);
-
-            return CreateWwwAuthenticateParameters(parameters);
+            return CreateFromWwwAuthenticationHeaderValue(wwwAuthenticateValue, string.Empty);
```
Doesn't this obsolete API only handle Bearer? Why do you pass string.empty here?

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -191,15 +295,144 @@ public static async Task<WwwAuthenticateParameters> CreateFromResourceResponseAs
+            }
+
+            //Special NTLM case does not have an a=b format
+            if (scheme == "NTLM")
```
We should not hardcode NTLM. Parser should be smart enough to handle case where some params are not in a=b format.

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -191,15 +295,144 @@ public static async Task<WwwAuthenticateParameters> CreateFromResourceResponseAs
+            {
+                return new WwwAuthenticateParameters
+                {
+                    RawParameters = new Dictionary<string, string>()
```
Maybe the dictionary does not hande this well. What does MSAL C++ do in this case? Should we have a `RawTokens` List<string> for params which are not in a=b format?

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -191,15 +295,144 @@ public static async Task<WwwAuthenticateParameters> CreateFromResourceResponseAs
+
+            var authValuesSplit = wwwAuthenticateValue.Split(new char[] { ' ' }, 2);
+
+            if (s_knownAuthenticationSchemes.Contains(authValuesSplit[0]))
```
NPE possible at `authValuesSplit[1] `

bgavrilMS replied:
The deprecated API `CreateFromWwwAuthenticationHeaderValue(wwwAuthenticateValue, string.Empty);` does not go into this flow :(

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -191,15 +295,144 @@ public static async Task<WwwAuthenticateParameters> CreateFromResourceResponseAs
+
+            IDictionary<string, string> parameters;
+
+            var authValuesSplit = wwwAuthenticateValue.Split(new char[] { ' ' }, 2);
```
I think there is an overload of Split that takes into account all whitespace chars.

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -191,15 +295,144 @@ public static async Task<WwwAuthenticateParameters> CreateFromResourceResponseAs
+
+            var authValuesSplit = wwwAuthenticateValue.Split(new char[] { ' ' }, 2);
+
+            if (s_knownAuthenticationSchemes.Contains(authValuesSplit[0]))
```
This is very confusing. `authValuesSplit[0]` should be the scheme, but instead you pass in the scheme as a param.

trwalke replied:
Discussed offline. leaving as is. This is a fix for the issue described here: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/3026

We are checking to see if the header value includes an auth scheme and are parsing it out. The old api may pass in a header value with a scheme attached when it should not. The new api will rely upon the authScheme parameter.

bgavrilMS commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -56,6 +58,77 @@ public void CreateWwwAuthenticateResponse(string resource, string authorizationU
         }
 
+        [TestMethod]
+        [DataRow("WLID1.0", "realm=WindowsLive, policy=MBI_SSL, siteId=\"ssl.live-tst.net\"")]
```
Can we have a test for each of the possible syntaxes in this spec? 

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate

Probably ok if we don't parse the single token. But let's make sure we don't crash.

pmaytak commented on src/client/Microsoft.Identity.Client/MsalErrorMessage.cs:
```diff
@@ -432,6 +432,9 @@ public static string CertMustHavePrivateKey(string certificateName)
 
         public const string RequestFailureErrorMessagePii = "=== Token Acquisition ({0}) failed:\n\tAuthority: {1}\n\tClientId: {2}.";
 
+        public const string UnableToParseAuthenticationHeader = "MSAL is unable to parse the authentication header returned from the resource endpoint. This can be a result of a malformed header returned in either the WWW-Authenticate or the Authentication-Info collections. See inner exception for details";
```
Add aka.ms link to pop wiki/docs?

pmaytak commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -58,6 +59,39 @@ public void CreateWwwAuthenticateResponse(string resource, string authorizationU
+            // Act
+            var authParams = WwwAuthenticateParameters.CreateFromAuthenticationHeaders(httpResponse.Headers, scheme);
+
+            // Assert
```
Add asserts? Maybe expected param count or something?

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -76,8 +95,59 @@ public class WwwAuthenticateParameters
+        /// <param name="cancellationToken">The cancellation token to cancel operation.</param>
+        /// <returns>WWW-Authenticate Parameters extracted from response to the unauthenticated call.</returns>
+        [System.ComponentModel.EditorBrowsable(EditorBrowsableState.Never)]
+        [Obsolete("This api is now obsolete and has been replaced with CreateFromAuthenticationResponseAsync(...)")]
```
```suggestion
        [Obsolete("This api is now obsolete and has been replaced with replaced with CreateFromAuthenticationResponseAsync(HttpResponseHeaders, string)")]
```

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -191,15 +300,129 @@ public static async Task<WwwAuthenticateParameters> CreateFromResourceResponseAs
+
             if (values.TryGetValue("authorization_uri", out value))
             {
                 wwwAuthenticateParameters.Authority = value.Replace("/oauth2/authorize", string.Empty);
```
Should we try to also replace `v2.0` from potential v2 auth endpoints like

`https://login.microsoftonline.com/common/v2.0/oauth2/authorize` ?

trwalke replied:
I'm not sure, is it expected that the `authorization_uri` to come with `v2.0` from the WWW-Auth params. @jmprieur ?

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -50,6 +58,17 @@ public class WwwAuthenticateParameters
+        /// <summary>
+        /// Pop Nonce. This is acquired from with the POP WWW-Authenticate header.
+        /// </summary>
+        public string PopNonce { get; private set; }
```
`PopNonce` implies that these WWW-Auth params are tied to the pop auth scheme.

pmaytak replied:
I thought you suggested to change this to PopNonce?

bgavrilMS replied:
Sorry, I did not explain this better. 

WWWAuthenticate header is associated with an auth scheme and each scheme can have a nonce. So there can be a nonce param for the POP WWWAuthhenticate header, but at the same time a nonce for the Bearer WWWAuthenticate header or for Basic etc. So in the context of `WWWAuthhenticate`, we cannot hardcode the name "pop" anywhere, as the header is tied to a scheme. It would be a bad api to have "POPNonce" property on the "Bearer" WWWAuth header.

However, the AuthenticationHeader class is a higher level parser. It aims to extract key information from a collection of WWWAuthenticate and from an AuthenticateInfo header. In higher level parser we can have `POPNonce`, and its meaning is "find the WWWAUThenticate header associated with POP scheme and extract the nonce value OR get the AUthenticateInfo nonce".

Hope this helps.

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -50,6 +58,17 @@ public class WwwAuthenticateParameters
+        public string AuthScheme { get; private set; }
+
+        /// <summary>
+        /// Pop Nonce. This is acquired from with the POP WWW-Authenticate header.
```
This is not true - it is acquired from any type of header which has the "nonce" param.

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -231,26 +454,12 @@ internal static WwwAuthenticateParameters CreateWwwAuthenticateParameters(IDicti
+            if (values.TryGetValue("nonce", out value))
             {
-                throw new ArgumentException(nameof(assignment), $"{assignment} isn't of the form a=b");
+                wwwAuthenticateParameters.PopNonce = value.Replace("\"", string.Empty);
```
Why do you replace " character?

trwalke replied:
I think this was some left-over logic from a previous implementation. removing

bgavrilMS commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -50,6 +58,17 @@ public class WwwAuthenticateParameters
+        public string PopNonce { get; private set; }
+
         /// <summary>
         /// Return the <see cref="RawParameters"/> of key <paramref name="key"/>.
```
I think we should exclude the resource / scope from the raw parameters as well. @jmprieur - thoughts?

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/WwwAuthenticateParametersIntegrationTests.cs:
```diff
@@ -60,5 +75,116 @@ public async Task CreateWwwAuthenticateResponseFromAzureResourceManagerUrlAsync(
             Assert.IsNull(authParams.Claims);
             Assert.AreEqual("invalid_token", authParams.Error);
```
might as well assert the RawParameters as well.

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/WwwAuthenticateParametersIntegrationTests.cs:
```diff
@@ -60,5 +75,116 @@ public async Task CreateWwwAuthenticateResponseFromAzureResourceManagerUrlAsync(
+
+            //Assert
+            Assert.IsTrue(parameterList.Any(param => param.AuthScheme == Constants.PoPAuthHeaderPrefix));
+            Assert.IsNotNull(parameterList.Where(param => param.AuthScheme == Constants.PoPAuthHeaderPrefix).Single().PopNonce);
```
nit: Simpler to use `enumerable.Single(condition)` instead of `enumerable.Where(condition).Single()`

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/WwwAuthenticateParametersIntegrationTests.cs:
```diff
@@ -60,5 +75,116 @@ public async Task CreateWwwAuthenticateResponseFromAzureResourceManagerUrlAsync(
+        }
+
+        [TestMethod]
+        public async Task ExtractNonceFromWwwAuthHeadersRawPamamsAsync()
```
Please combine it with the previous test. Integration tests are expensive time wise, better to minimize their number.

bgavrilMS commented on src/client/Microsoft.Identity.Client/AuthenticationHeaderParser.cs:
```diff
@@ -0,0 +1,158 @@
+        {
+            AuthenticationHeaderParser authenticationHeaderParser = new AuthenticationHeaderParser();
+            AuthenticationInfoParameters authenticationInfoParameters = new AuthenticationInfoParameters();
+            string serverNonce = null;
```
not needed.

bgavrilMS commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -56,6 +59,86 @@ public void CreateWwwAuthenticateResponse(string resource, string authorizationU
+            var authParams = WwwAuthenticateParameters.CreateFromAuthenticationHeaders(httpResponse.Headers, scheme);
+
+            // Assert
+            //if (scheme == "NTLM")
```
dead code?

bgavrilMS replied:
This test doesn't assert anything.

bgavrilMS commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -56,6 +59,86 @@ public void CreateWwwAuthenticateResponse(string resource, string authorizationU
+        [TestMethod]
+        [DataRow("AuthScheme", "" )]
+        [DataRow("AuthScheme", "realm = someRealm")]
+        [DataRow("AuthScheme", "token68")]
```
I ran this test and I am not sure this is the correct result? The `RawParameters` has 1 entry { "AuthScheme": "token68" } which doesn't seem correct? 

I'd would think you either add this param in `RawParameters` as { "token68": null } or {"token68" : "token68"} 

@pmaytak, since you also reviewed this - thoughts?

trwalke replied:
Updated logic here.

pmaytak replied:
Is token68 an actual type of a token...? If so `{"token68" : "ActualToken68Token"}` makes sense. I'll take another look.

bgavrilMS commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -56,6 +59,86 @@ public void CreateWwwAuthenticateResponse(string resource, string authorizationU
+        [DataRow("AuthScheme", "" )]
+        [DataRow("AuthScheme", "realm = someRealm")]
+        [DataRow("AuthScheme", "token68")]
+        [DataRow("AuthScheme", "realm = someRealm token68")]
```
This also looks wrong. I'd expect RawParameters to have `{"realm":"someRealm"}` ?

<img width="303" alt="image" src="https://user-images.githubusercontent.com/12273384/204033601-cce18a56-811c-4c3e-91ba-03052daa8606.png">

bgavrilMS commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -56,6 +59,86 @@ public void CreateWwwAuthenticateResponse(string resource, string authorizationU
+        [DataRow("AuthScheme", "token68")]
+        [DataRow("AuthScheme", "realm = someRealm token68")]
+        [DataRow("AuthScheme", "auth-param1=token1, auth-param2=token2, auth-param3=token3")]
+        [DataRow("AuthScheme", "realm = someRealm auth-param1=token1, auth-param2=token2, auth-param3=token3")]
```
This is wrong: 

<img width="521" alt="image" src="https://user-images.githubusercontent.com/12273384/204033742-45a31d59-70ab-44dd-a6be-42eed13f1dfe.png">

bgavrilMS commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -56,6 +59,86 @@ public void CreateWwwAuthenticateResponse(string resource, string authorizationU
             Assert.IsNull(authParams.Error);
         }
 
+        [TestMethod]
```
This test shows that if "token68" is a param, the parser is broken.

bgavrilMS commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -56,6 +59,86 @@ public void CreateWwwAuthenticateResponse(string resource, string authorizationU
+
+        [TestMethod]
+        [DataRow("WLID1.0", "realm=WindowsLive, policy=MBI_SSL, siteId=\"ssl.live-tst.net\"")]
+        [DataRow("NTLM", "dG9rZW42OA==")]
```
As before, RawParameters should probably not have {"NTLM": "dG9.."} as it does not reflect the state.

bgavrilMS commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -56,6 +59,86 @@ public void CreateWwwAuthenticateResponse(string resource, string authorizationU
+            if (scheme == "NTLM")
+            {
+                Assert.AreEqual(scheme, authParams.AuthScheme);
+                Assert.AreEqual(values, authParams.RawParameters[scheme]);
```
nit: better to use actual strings in the assert.

bgavrilMS commented on src/client/Microsoft.Identity.Client/MsalErrorMessage.cs:
```diff
@@ -432,6 +432,9 @@ public static string CertMustHavePrivateKey(string certificateName)
 
         public const string RequestFailureErrorMessagePii = "=== Token Acquisition ({0}) failed:\n\tAuthority: {1}\n\tClientId: {2}.";
 
+        public const string UnableToParseAuthenticationHeader = "MSAL is unable to parse the authentication header returned from the resource endpoint. This can be a result of a malformed header returned in either the WWW-Authenticate or the Authentication-Info collections. See inner exception for details. For more information on Proof-of-Possession, please see  https://aka.ms/msal-net-pop";
```
I would not say anything about POP in this message.

bgavrilMS commented on src/client/Microsoft.Identity.Client/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,90 @@
+                    return parameters;
+
+                    //Could not get Auth info parameters
+                    throw new MsalClientException(MsalError.UnableToParseAuthenticationHeader, MsalErrorMessage.UnableToParseAuthenticationHeader);
```
The ex message asks the user to look at the inner exception. But in this case there is no inner exception. Can you please add a custom error message. Have it in the code right here, to make the code easier to read.

bgavrilMS replied:
The exception messages should contain the header value, for supportability.

bgavrilMS replied:
If header is not present, you should behave like WWW-Authenticate parser, which doesn't throw. Note: this is a blocking comment

bgavrilMS commented on src/client/Microsoft.Identity.Client/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,90 @@
+
+            try
+            {
+                if (httpResponseHeaders.Contains(AuthenticationInfoKey))
```
nit: 

var authInfoHeader = httpResponseHeaders.SingleOrDefault(AuthenticationInfoKey);

if (authInfoHeader != null) { .... }

bgavrilMS commented on src/client/Microsoft.Identity.Client/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,90 @@
+            }
+            catch (Exception ex) when (ex is not MsalClientException)
+            {
+                throw new MsalClientException(MsalError.UnableToParseAuthenticationHeader, MsalErrorMessage.UnableToParseAuthenticationHeader, ex);
```
This is inconsistent with WWW-Auth parser, which returns an empty object in case there is no header. This is a blocking comment.

bgavrilMS commented on src/client/Microsoft.Identity.Client/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,90 @@
+            {
+                if (httpResponseHeaders.Contains(AuthenticationInfoKey))
+                {
+                    var authInfoValue = httpResponseHeaders.Where(header => header.Key == AuthenticationInfoKey).Single().Value.FirstOrDefault();
```
Single() can throw an InvalidOpException if there are 2+ elements. Better to use `Where()` and throw specific exceptions if there is 0 or 2+ elements.

trwalke replied:
I tried to repro this and it is not possible for the HttpResponseHeaders to contain 2 of the same header. It will simply create an array of all of the values assigned to the same key. It still represents one header.

bgavrilMS commented on src/client/Microsoft.Identity.Client/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,90 @@
+                    var authInfoValue = httpResponseHeaders.Where(header => header.Key == AuthenticationInfoKey).Single().Value.FirstOrDefault();
+
+
+                    var AuthValuesSplit = authInfoValue.Split(new char[] { ' ' }, 2);
```
1. naming
2. The parsing code is not defensive enough, and relies on try / catch too much. As a consequence, you throw exception which are generic and hard to root cause. 

I suggest you try to throw more exceptions, e.g. if you split by ' ' and `AuthValuesSplit[1]` doesn't exist, you need to throw proper exception.

if

bgavrilMS reviewed (changes requested):
1. "token68" completely throws your parser off
2. WWWAuth and AuthInfo parsers behave differently when the header is not there
3. POPNonce is a bad name for a generic param
4. Exceptions are not specific enough. 

Have you considered abandoning the custom parsing logic you have and adopting the MSAL C++ parser instead?

bgavrilMS commented on src/client/Microsoft.Identity.Client/AuthenticationHeaderParser.cs:
```diff
@@ -0,0 +1,158 @@
+    /// <summary>
+    /// Parsed authentication headers to retrieve header values from HttpResponseHeaders.
+    /// </summary>
+    public class AuthenticationHeaderParser
```
Since the public API is stablized, could you please create a class diagram or a table with all the public API in these classes, ensuring that you mark: 

- what API is old
- what API is new
- what API has been deprecated and what we're suggesting to replace it with

This will make it easier to review this doc and should be added to the API Review spec as well.

trwalke replied:
Updated PR description

pmaytak commented on src/client/Microsoft.Identity.Client/WwwAuthenticateParameters.cs:
```diff
@@ -231,26 +462,12 @@ internal static WwwAuthenticateParameters CreateWwwAuthenticateParameters(IDicti
-                .ToArray();
-
-            if (segments.Length != 2)
+            if (values.TryGetValue("nonce", out value))
```
Do we have test cases where the parameter is not lowercase, for ex:
`        [DataRow("AuthScheme", "realm=someRealm Nonce=nonce, ERROR=\"err\"")]`

pmaytak commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -56,6 +59,129 @@ public void CreateWwwAuthenticateResponse(string resource, string authorizationU
+            }
+        }
+
+        [TestMethod]
```
Does the current implementation support these scenarios?
- [a response can have multiple challenges with the same auth-scheme but with different realms.](https://httpwg.org/specs/rfc9110.html#rfc.section.11.5)
```csharp
httpResponse.Headers.Add(WwwAuthenticateHeaderName, "Bearer realm=\"first\", Bearer realm=\"second\"");
```
```csharp
httpResponse.Headers.Add(WwwAuthenticateHeaderName, "Bearer realm=\"first\"");
httpResponse.Headers.Add(WwwAuthenticateHeaderName, "Bearer realm=\"second\"");
 ```
- Header field [might contain more than one challenge](https://httpwg.org/specs/rfc9110.html#rfc.section.11.6.1)
```csharp
httpResponse.Headers.Add(WwwAuthenticateHeaderName, "Bearer realm=\"someRealm\", OtherScheme realm=\"someRealm\"");
```
```csharp
httpResponse.Headers.Add(WwwAuthenticateHeaderName, "Bearer realm=\"someRealm\"");
httpResponse.Headers.Add(WwwAuthenticateHeaderName, "OtherScheme realm=\"someRealm\"");
 ```

pmaytak commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -56,6 +59,129 @@ public void CreateWwwAuthenticateResponse(string resource, string authorizationU
+        [DataRow("AuthScheme", "realm=someRealm")]
+        [DataRow("AuthScheme", "realm=someRealm token68")]
+        [DataRow("AuthScheme", "realm=someRealm auth-param1=token1, auth-param2=token2, auth-param3=token3")]
+        [DataRow("AuthScheme", "realm=someRealm token68 auth-param1=token1, auth-param2=token2, auth-param3=token3")]
```
I am not clear from the specs - what is the separator?
[This HTTP spec](https://httpwg.org/specs/rfc9110.html#rfc.section.11.6.1) says 
> each challenge can contain a comma-separated list of authentication parameters

And same with [different section](https://httpwg.org/specs/rfc9110.html#rfc.section.11.2) in the same spec
> The authentication scheme is followed by additional information necessary for achieving authentication via that scheme as either a comma-separated list of parameters or a single sequence of characters capable of holding base64-encoded information.

Mozilla has examples of both - [comma separated](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate#basic_authentication) and [with spaces](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate#syntax) after realm and token68.

This test case passes
`[DataRow("AuthScheme", "realm=someRealm, auth-param1=token1, auth-param2=token2, auth-param3=token3")]`
but this does not
`[DataRow("AuthScheme", "realm=someRealm, token68, auth-param1=token1, auth-param2=token2, auth-param3=token3")]`

trwalke replied:
I think I found a bug in our `SplitWithQuotes` methods that seems to fail when spaces are included within the values. Ill see if I can resolve it.

trwalke replied:
Ok, there was no bug there, I found the issue in the test code. 
In regards to the seperator, there doesn't seem to be a clear indicator of what it is in the spec. It seems like it can either be a space or a comma followed by a space. So, we are accounting for both of these scenarios.

pmaytak replied:
Thanks for verifying. Let's make sure we have test cases that cover these scenarios to protect from regressions.

trwalke commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -56,6 +59,129 @@ public void CreateWwwAuthenticateResponse(string resource, string authorizationU
+        }
+
+        [TestMethod]
+        [DataRow("WLID1.0", "realm=WindowsLive, policy=MBI_SSL, siteId=\"ssl.live-tst.net\"")]
```
Is this a valid format?

bgavrilMS commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -56,6 +59,137 @@ public void CreateWwwAuthenticateResponse(string resource, string authorizationU
+            // Assert
+            Assert.AreEqual(scheme, authParams.AuthScheme);
+
+            if (string.IsNullOrEmpty(values))
```
You could benefit from having separate tests - this assertion block is difficult to read and could miss important details. Better to a have a few different tests or use something like below to differentiate:

```csharp
[DataRow("AuthScheme1", "")]
[DataRow("AuthScheme2", "token68")]
```

bgavrilMS commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -56,6 +59,137 @@ public void CreateWwwAuthenticateResponse(string resource, string authorizationU
+            var authParams = WwwAuthenticateParameters.CreateFromAuthenticationHeaders(httpResponse.Headers, scheme);
+
+            // Assert
+            if (scheme == "NTLM")
```
These asserts are easier to read.

bgavrilMS commented on src/client/Microsoft.Identity.Client/AuthenticationInfoParameters.cs:
```diff
@@ -0,0 +1,95 @@
+            }
+            catch (Exception ex) when (ex is not MsalClientException)
+            {
+                throw new MsalClientException(MsalError.UnableToParseAuthenticationHeader, MsalErrorMessage.UnableToParseAuthenticationHeader + $" See inner exception for details.", ex);
```
Please add the header value to the exception message.

bgavrilMS commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -205,6 +364,14 @@ public async Task CreateFromResourceResponseAsync_HttpClient_Null_Async(HttpClie
             Func<Task> action = () => WwwAuthenticateParameters.CreateFromResourceResponseAsync(httpClient, resourceUri);
 
             await Assert.ThrowsExceptionAsync<ArgumentNullException>(action).ConfigureAwait(false);
+
```
This test should not be a data test, it has only 1 "row"

bgavrilMS reviewed (approved):
Aproving, but please simplify public API by allowing only methods with default cancellation token (see comment)

pmaytak commented on tests/Microsoft.Identity.Test.Unit/WwwAuthenticateParametersTests.cs:
```diff
@@ -598,6 +650,32 @@ private static HttpResponseMessage CreateBearerAndPopHttpResponse(bool combinedC
+            {
+                Headers =
+                {
+                    { WwwAuthenticateHeaderName, $"Bearer AUTHORIZATION_URI=\"https://login.microsoftonline.com/TenantId/oauth2/authorize\"" },
```
Can you also make the scheme name be uppercase, I think that's valid scenario?

trwalke replied:
Cant do that for the scheme, it is case sensitive
